File: src/checks/resilience.ts
import { Check, CheckContext, CheckResult } from '../types.js';
import { createClient, DripError } from '../drip-client.js';

export const getMetricsCheck: Check = {
  name: 'SDK Metrics',
  description: 'Get SDK metrics (requires resilience enabled)',
  quick: true,
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);

    try {
      const sdk = (client as unknown as {
        sdk: {
          getMetrics?: () => Promise<{
            requests: { total: number; successful: number; failed: number };
            latency: { average: number; p95: number; p99: number };
          }>;
        };
      }).sdk;

      if (!sdk.getMetrics) {
        const duration = performance.now() - start;
        return {
          name: 'SDK Metrics',
          success: true,
          duration,
          message: 'Skipped (getMetrics not available)',
          details: 'The getMetrics method is not available in the SDK. May require resilience to be enabled.',
        };
      }

      const result = await sdk.getMetrics();
      const duration = performance.now() - start;

      return {
        name: 'SDK Metrics',
        success: true,
        duration,
        message: `Requests: ${result.requests.total} (${result.requests.successful} ok)`,
        details: `Avg latency: ${result.latency.average}ms, P95: ${result.latency.p95}ms`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'SDK Metrics',
          success: true,
          duration,
          message: 'Skipped (resilience not enabled)',
          details: 'Enable resilience in SDK config to access metrics',
        };
      }

      return {
        name: 'SDK Metrics',
        success: false,
        duration,
        message: err.message || 'Failed to get metrics',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

export const getHealthCheck: Check = {
  name: 'Resilience Health',
  description: 'Get resilience health status',
  quick: true,
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);

    try {
      const sdk = (client as unknown as {
        sdk: {
          getHealth?: () => Promise<{
            status: 'healthy' | 'degraded' | 'unhealthy';
            circuitBreaker: { state: string; failures: number };
            rateLimit: { remaining: number; resetAt: string };
          }>;
        };
      }).sdk;

      if (!sdk.getHealth) {
        const duration = performance.now() - start;
        return {
          name: 'Resilience Health',
          success: true,
          duration,
          message: 'Skipped (getHealth not available)',
          details: 'The getHealth method is not available in the SDK. May require resilience to be enabled.',
        };
      }

      const result = await sdk.getHealth();
      const duration = performance.now() - start;

      return {
        name: 'Resilience Health',
        success: true,
        duration,
        message: `Status: ${result.status}`,
        details: `Circuit: ${result.circuitBreaker.state}, Rate limit remaining: ${result.rateLimit.remaining}`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Resilience Health',
          success: true,
          duration,
          message: 'Skipped (resilience not enabled)',
          details: 'Enable resilience in SDK config to access health status',
        };
      }

      return {
        name: 'Resilience Health',
        success: false,
        duration,
        message: err.message || 'Failed to get health status',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

File: python/checks/resilience.py
"""Resilience and metrics checks."""
from ..types import Check, CheckContext, CheckResult
from ..drip_client import create_client


async def _get_metrics_check(ctx: CheckContext) -> CheckResult:
    """Get SDK metrics."""
    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'get_metrics'):
            return CheckResult(
                name="sdk_metrics",
                success=True,
                duration=0,
                message="Skipped (get_metrics not available)",
                details="The get_metrics method is not available in the SDK"
            )

        metrics = client.get_metrics()

        if metrics is None:
            return CheckResult(
                name="sdk_metrics",
                success=True,
                duration=0,
                message="Skipped (resilience not enabled)",
                details="Enable resilience to get metrics"
            )

        total_requests = metrics.get('total_requests', 0) if isinstance(metrics, dict) else getattr(metrics, 'total_requests', 0)

        return CheckResult(
            name="sdk_metrics",
            success=True,
            duration=0,
            message=f"Total requests: {total_requests}",
            details="Metrics retrieved successfully"
        )
    except Exception as e:
        error_str = str(e)
        if '404' in error_str or '501' in error_str:
            return CheckResult(
                name="sdk_metrics",
                success=True,
                duration=0,
                message="Skipped (endpoint not implemented)",
                details=error_str
            )
        return CheckResult(
            name="sdk_metrics",
            success=False,
            duration=0,
            message=f"Failed to get metrics: {e}"
        )


get_metrics_check = Check(
    name="sdk_metrics",
    description="Get SDK metrics",
    run=_get_metrics_check
)


async def _get_health_check(ctx: CheckContext) -> CheckResult:
    """Get resilience health status."""
    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'get_health'):
            return CheckResult(
                name="resilience_health",
                success=True,
                duration=0,
                message="Skipped (get_health not available)",
                details="The get_health method is not available in the SDK"
            )

        health = client.get_health()

        if health is None:
            return CheckResult(
                name="resilience_health",
                success=True,
                duration=0,
                message="Skipped (resilience not enabled)",
                details="Enable resilience to get health status"
            )

        if isinstance(health, dict):
            status = health.get('healthy', health.get('status', 'unknown'))
        else:
            status = getattr(health, 'healthy', getattr(health, 'status', 'unknown'))

        return CheckResult(
            name="resilience_health",
            success=True,
            duration=0,
            message=f"Health status: {status}",
            details="Health check completed"
        )
    except Exception as e:
        error_str = str(e)
        if '404' in error_str or '501' in error_str:
            return CheckResult(
                name="resilience_health",
                success=True,
                duration=0,
                message="Skipped (endpoint not implemented)",
                details=error_str
            )
        return CheckResult(
            name="resilience_health",
            success=False,
            duration=0,
            message=f"Failed to get health: {e}"
        )


get_health_check = Check(
    name="resilience_health",
    description="Get resilience health status",
    run=_get_health_check
)
